### PR TITLE
Changing iree/base/math.h to C and adding some PRNG routines.

### PR DESCRIPTION
**Pull request by benvanik**
These routines will be used for threadpool worker theft distribution where it's important that they are relatively fast.

I used this opportunity to clean up the file and finish some of the TODOs for when C happened.

### CODE REVIEW

File: iree/base/math.h [hunk 1]
```diff
@@ -117,99 +129,267 @@ IREE_FORCEINLINE int CountLeadingZeros64(uint64_t n) {
...
+  n |= n >> 4;
+  n |= n >> 8;
+  n |= n >> 16;
+  n |= n >> 32;
+  return n + 1;
+#endif  // 1
+}
+
+//==============================================================================
+// Pseudo-random number generators (PRNGs): **NOT CRYPTOGRAPHICALLY SECURE*
```
Comment (bjacob) [inline iree/base/math.h:271]: this is a bit of an understatement. more descriptive would be "very cheap PRNGs, high performance, low quality".

The c++ `<random>` header defines some useful PRNGs. If we're building our own here, that means we know something specific to our intended applications that requires different PRNGs. What is it?

(The default PRNG to beat is linear congruential, since that's a single instruction, SIMDs trivially, is clearly uniformly distributed over the whole integer type, and is readily available as `std::minstd_rand`).

File: iree/base/math_test.cc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+#include "iree/testing/status_matchers.h"
+
+namespace {
+
+//==============================================================================
+// Bitwise rotation (aka circular shifts)
+//==============================================================================
+
+TEST(BitwiseRotationTest, ROTL64) {
+  EXPECT_EQ(0ull, iree_math_rotl_u64(0ull, 0u));
```
Comment (bjacob) [inline iree/base/math_test.cc:27]: Could we have a bit more coverage? Maybe some loop, as in some of the other tests below, to cover more cases with less code.

This isn't to be an insufferable reviewer, but I'm guessing that this is a kind of code that various people will contribute fast/crazy/hardware-specific implementations to in the future, so you want to feel able to take their PRs without doing too much testing mentally everytime ;-)

[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math_test.cc:27]: you often do with EXPECT_EQ - otherwise 1 is a signed integer and you'll get type coercion warnings. 0 is a somewhat special case, but I'd rather be consistent :)

File: iree/base/math.h [hunk 3]
```diff
@@ -26,88 +28,98 @@
...
+// Bitwise rotation (aka circular shifts)
+//==============================================================================
+
+// Unsigned rotate-left a 64-bit integer.
+// https://en.cppreference.com/w/cpp/numeric/rotl
+//
+//
+// NOTE: this exact form is confirmed to be recognized by the compilers we care
+// about; do not modify: https://godbolt.org/z/xzof9d
+static inline uint64_t iree_math_rotl_u64(const uint64_t n, uint32_t c) {
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math.h:51]: Desired; `static inline` is safe to put in header files; the static ensures that if it is not inlined it'll be emitted into a compilation unit just as if static had been used there making it unique to that complication unit and avoiding ODR violations.
See: https://www.greenend.org.uk/rjk/tech/inline.html
TLDR: "A function defined with static inline. Stand-alone object code may be emitted if required. You can have multiple definitions in your program, in different translation units, and it will still work."

File: iree/base/math.h [hunk 4]
```diff
@@ -26,88 +28,98 @@
...
-#elif defined(__GNUC__)
-  // Use __builtin_clz, which uses the following instructions:
-  //  x86: bsr
-  //  ARM64: clz
-  //  PPC: cntlzd
-  static_assert(sizeof(int) == sizeof(n),
-                "__builtin_clz does not take 32-bit arg");
-
+#elif defined(IREE_COMPILER_GCC_COMPAT)
 #if defined(__LCZNT__)
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math.h:82]: see above: each compilation unit emitting the function - if it could not be inlined - will have its own copy

this is also an internal IREE file and not usable from non-IREE code and all our compilation units will have the same settings

File: iree/base/math.h [hunk 5]
```diff
@@ -117,99 +129,267 @@ IREE_FORCEINLINE int CountLeadingZeros64(uint64_t n) {
...
+//
+// Which to choose is entirely determined by fiddling on godbolt for the
+// target platform: https://godbolt.org/z/h4vPzo
+
 // Rounds up the value to the nearest power of 2 (if not already a power of 2).
-IREE_FORCEINLINE int RoundUpToNearestPow2(int n) {
-  return n ? ~0u >> LeadingZeros(n) : 1;
+// For 32-bit numbers this only supports values <= 2^31; higher will wrap.
+static inline uint32_t iree_math_round_up_to_pow2_u32(uint32_t n) {
+#if 0    // golf required; can be bloated
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math.h:231]: it's effectively a comment that can be turned into executable code with a single character swap for easy hacking (in this case, all bodies of the if-else have identical behavior and are just differing potential implementations)

File: iree/base/math.h [hunk 6]
```diff
@@ -117,99 +129,267 @@ IREE_FORCEINLINE int CountLeadingZeros64(uint64_t n) {
...
+//==============================================================================
+
+// A fixed-increment version of Java 8's SplittableRandom generator
+// See http://dx.doi.org/10.1145/2714064.2660195 and
+// http://docs.oracle.com/javase/8/docs/api/java/util/SplittableRandom.html
+//
+// SplitMix64 as recommended for use with xoroshiro by the authors:
+// http://prng.di.unimi.it/splitmix64.c
+// http://rosettacode.org/wiki/Pseudo-random_numbers/Splitmix64
+typedef uint64_t iree_prng_splitmix64_state_t;
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math.h:281]: this file is for use in C code, your fancy modern C++ stuff does not apply ;P

File: iree/base/math.h (hunk 1 above)
Comment (benvanik) [reply iree/base/math.h:271]: This is for C (so no <random>) and a very specific set of use cases where code-size and cost/iteration matters; the goal is to have something that gives an adequate distribution with commensurate cost: in this case, 16 bytes of state and 5-6 instructions per generation. This particular PRNG was designed for the use cases we are applying it to and is used in quite a few schedulers (including the java, mesa, solaris, and dtrace thread schedulers). It's also used quite a bit more broadly in things like numba and numpy. You don't have the follow on PR for context but it's literally used to select a number between 0-63 to bias a work-stealing queue - and even then this may be _too good_ for that use case ;)

File: iree/base/math.h [hunk 7]
```diff
@@ -26,88 +28,98 @@
...
+//==============================================================================
 
-IREE_FORCEINLINE int CountLeadingZeros32(uint32_t n) {
-#if defined(_MSC_VER)
+static inline int iree_math_count_leading_zeros_u32(const uint32_t n) {
+#if defined(IREE_COMPILER_MSVC)
   unsigned long result = 0;  // NOLINT(runtime/int)
   if (_BitScanReverse(&result, n)) {
-    return 31 - result;
+    return (int)(31 - result);
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/base/math.h:78]: This is C, so C casts are all there is.

File: iree/base/math_test.cc (hunk 2 above)
Comment (benvanik) [reply iree/base/math_test.cc:27]: More coverage would be great yes - but there was literally _zero_ coverage before and now there is _some_ that I added while touching this code as part of an incremental code health improvement :)

File: iree/base/math.h (hunk 1 above)
Comment (bjacob) [reply iree/base/math.h:271]: So, for such low-quality 6-bit valus, a 8bit LCG gives you a ratio of 1 instruction (intrinsic = vmulq_u8) for 16 generated random bytes on ARM NEON.  Can you beat that? :-)

**Review by bjacob — APPROVED** [review]
thanks for all the answers! still kind of interested in why no 8bit LCG, and any additional testing.

File: iree/base/math.h (hunk 1 above)
Comment (benvanik) [reply iree/base/math.h:271]: oo neat, didn't know that existed! that's probably worth using for that particular worker selection case!
(after I land threading, I'm not looking to golf any more until it's in ;)

**Comment by benvanik** [comment]
> and any additional testing.

if you can think of useful test cases please add - for some of the tests in here I was literally scrounging on sourcegraph for what other major projects (linux, dart, qemu, etc) tested their stuff with and came up short.
